File: 3/3000s/3396/MinimumNumberOfOperationsToMakeElementsInArrayDistinct.java
/*
https://leetcode.com/problems/minimum-number-of-operations-to-make-elements-in-array-distinct/description/?envType=daily-question&envId=2025-04-08
*/

class Solution {

    public static boolean isValidArray(int[] nums){
        int lengthOfNums=nums.length;
        if(lengthOfNums==0) return true;
        HashMap<Integer, Integer>numsItemsMap=new HashMap<>();
        for(int i=0; i<lengthOfNums; ++i){
            if(numsItemsMap.containsKey(nums[i])){
                return false;
            }
            numsItemsMap.put(nums[i], nums[i]);
        }
        return true;
    }

    public int minimumOperations(int[] nums) {
        int resultantMinOperations=0;
        while(true){
            if(isValidArray(nums)) break;
            if(nums.length<3){
                resultantMinOperations++;
                break;
            } else{
                int[] tempNums=new int[nums.length-3];
                for(int i=3; i<nums.length; ++i){
                    tempNums[i-3]=nums[i];
                }
                resultantMinOperations++;
                nums=tempNums;
            }
        }
        return resultantMinOperations;
    }
}

File: 3/3000s/3396/MinimumNumberOfOperationsToMakeElementsInArrayDistinct.cpp
/*
https://leetcode.com/problems/minimum-number-of-operations-to-make-elements-in-array-distinct/description/?envType=daily-question&envId=2025-04-08
*/

class Solution {
public:

    bool isValid(vector<int>&nums){
        int sz=nums.size();
        if(sz==0) return true;
        unordered_map<int, int>mp;
        for(int i=0; i<sz; ++i){
            if(mp.find(nums[i])!=mp.end()) return false;
            mp[nums[i]]=nums[i];
        }
        return true;
    }

    int minimumOperations(vector<int>& nums) {
        int sz=nums.size(), res=0;
        while(true){
            if(isValid(nums)) break;
            if(nums.size()<3){
                res++;
                break;
            } else{
                nums[0]=-1, nums[1]=-1, nums[2]=-1;
                nums.erase(remove(nums.begin(), nums.end(), -1), nums.end());
                res++;
            }
        }
        return res;
    }
};
